byterun/domain.c

```diff
@@ -324,16 +366,43 @@ CAMLprim value caml_domain_spawn(value callback)
   if (p.newdom) {
     /* successfully created a domain.
        p.callback is now owned by that domain */
-    pthread_detach(th);
   } else {
     /* failed */
-    void* r;
-    pthread_join(th, &r);
+    caml_ml_domain_join(th_val);
     caml_delete_root(p.callback);
     caml_failwith("failed to allocate domain");
   }
 
-  CAMLreturn (Val_unit);
+  CAMLreturn (th_val);
+}
+
+CAMLprim value caml_ml_domain_join(value domain)
+{
+    CAMLparam1 (domain);
+    struct domain_thread **dp = Domainthreadptr_val(domain);
+    struct domain_thread *d;
+    int err = 0;
+    pthread_t th;
+
+    Assert(dp);
+    d = *dp;
+    Assert(d);
+    th = d->th;
```

Comment:
I've added a comment, I think it is clearer that way. Otherwise it could be confusing why copying the `pthread_t` needed the lock, but operating on it (with `pthread_join`) didn't.